### PR TITLE
 Text coloring based on topic model, support category column

### PR DESCRIPTION
**Pull request by hideaki**
# Description
- Text coloring based on topic model
- Support category column

# Checklist
Make sure you have performed the following items before submitting this pull request.
If not, please describe the reason.  

- [x] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/textanal.R [hunk 1]
```diff
@@ -293,6 +306,19 @@ tidy.textanal_exploratory <- function(x, type="word_count", max_words=NULL, max_
...
     }
+
+    # If there is category_col, create data frame whose row represents a category-word combination, for bar chart with category color.
+    if (type == "category_word_count" && !is.null(x$category_col)) {
+      res2 <- dfm_to_df(x$dfm)
+      if (!is.null(max_words)) { # filter with the top words.
+        res2 <- res2 %>% filter(token %in% res$word)
+      }
+      # Join document info.
+      res2 <- res2 %>% left_join(x$df %>% select(!!rlang::sym(x$category_col)) %>% mutate(doc_id=row_number()), by=c(document="doc_id")) %>%
```
Comment (kei51e) [inline R/textanal.R:317]: Add namespace to mutate and select.

File: R/textanal.R [hunk 2]
```diff
@@ -631,19 +685,62 @@ tidy.textmodel_lda_exploratory <- function(x, type = "doc_topics", num_top_words
...
-    docs_coordinates_df <- as.data.frame(x$docs_coordinates)
-    res <- res %>% dplyr::bind_cols(docs_coordinates_df)
+    res <- x$doc_df
+  }
+  else if (type == "doc_topics_tagged") {
+    words_to_tag_df <- x$doc_word_df %>% distinct(document, word, .keep_all = TRUE)
+    words_to_tag_df <- words_to_tag_df %>% dplyr::mutate(max_topic = summarize_row(across(starts_with("topic")), which.max.safe), topic_max = summarize_row(across(starts_with("topic")), max))
+    words_to_tag_df <- words_to_tag_df %>% dplyr::group_by(document) %>% dplyr::slice_max(topic_max, prop=0.3) %>% dplyr::ungroup() # Filter per document.
+    tag_df <- words_to_tag_df %>% dplyr::nest_by(document) %>% ungroup()
+    res <- x$doc_df %>% rename(text=!!x$text_col) %>% mutate(doc_id=row_number()) %>% left_join(tag_df, by=c("doc_id"="document"))
```
Comment (kei51e) [inline R/textanal.R:695]: Add namespace to distinct, mutate, ungroup, rename and select.

File: R/textanal.R [hunk 3]
```diff
@@ -631,19 +685,62 @@ tidy.textmodel_lda_exploratory <- function(x, type = "doc_topics", num_top_words
...
+    res <- res %>% mutate(tagged_text=purrr::flatten_chr(purrr::map2(text, data, function(txt,dat) {
+      if (!is.null(dat)) {
+        orig_strs <- list()
+        for (i in 1:nrow(dat)) {
+          if (stringr::str_detect(dat$word[i], '[a-zA-Z]')) { # For alphabet word, char before/after should not be alphabet, to avoid matches within other words.
+            pre_regex <- '(?<![a-zA-Z])\\Q'
```
Comment (kei51e) [inline R/textanal.R:701]: Explain the expression.

File: R/textanal.R [hunk 4]
```diff
@@ -631,19 +685,62 @@ tidy.textmodel_lda_exploratory <- function(x, type = "doc_topics", num_top_words
...
+            post_regex <- '\\E(?![a-zA-Z])'
+          }
+          else {
+            pre_regex <- '\\Q'
+            post_regex <- '\\E'
+          }
+          orig_strs_ <- stringr::str_extract_all(txt, stringr::regex(stringr::str_c(pre_regex, dat$word[i], post_regex), ignore_case = TRUE))
+          orig_strs[[i]] <- orig_strs_[[1]]
+          txt <- stringr::str_replace_all(txt, stringr::regex(stringr::str_c(pre_regex, dat$word[i], post_regex), ignore_case = TRUE), stringr::str_c('_____', i, '_____'))
```
Comment (kei51e) [inline R/textanal.R:710]: It will break if the word is a number. For example `_____2_____` matches the word "2".

File: R/textanal.R (hunk 2 above)
Comment (hideaki) [reply R/textanal.R:695]: Done.

File: R/textanal.R (hunk 4 above)
Comment (hideaki) [reply R/textanal.R:711]: Will track it as an issue.

**Pull request merged**